### PR TITLE
Feature/hard checkpoints issue 438

### LINKED ISSUE #438: [POS] Implement checkpointing similarly to StratisX
these are long term checkpoints that make IBD much faster as the block validation takes a lot of time

### PR DESCRIPTION
**Pull request by Aprogiena**
Implements universal (for both Bitcoin and Stratis) hardcoded checkpoints. This PR, however, only speed up Stratis blockchain. Bitcoin checkpoints are defined but not used in this PR.

Stratis block validation is now skipped including stake modifier calculations for all blocks under the last checkpoint.

This PR also fixes a freshly found bug in `StakeValidator.ComputeNextStakeModifier`.

Testnet with 119k blocks fully synced with this update in 14 minutes (last checkpoint at block 115k).

Mainnet with 530k blocks in 4h 30m (kast checkpoint at block 528k). See #477 for more.

Fix #438

### CODE REVIEW

File: Stratis.Bitcoin.Features.Consensus/Checkpoints.cs [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    /// Description of checkpointed block.
+    /// </summary>
+    public class CheckpointInfo
+    {
+        /// <summary>Hash of the checkpointed block header.</summary>
+        public uint256 Hash { get; set; }
+
+        /// <summary>Stake modifier value of the checkpointed block.</summary>
+        /// <remarks>Stratis only.</remarks>
+        public ulong StakeModifier { get; set; }
```
Comment (dangershony) [inline Stratis.Bitcoin.Features.Consensus/Checkpoints.cs]: Drop this for now I think its not used

File: Stratis.Bitcoin.Features.Consensus/ConsensusFeature.cs [hunk 2]
```diff
@@ -223,6 +223,7 @@ public static IFullNodeBuilder UseConsensus(this IFullNodeBuilder fullNodeBuilde
                     // TODO: this should be set on the network build
                     fullNodeBuilder.Network.Consensus.Options = new PowConsensusOptions();
 
+                    services.AddSingleton<ICheckpoints>(new Checkpoints(fullNodeBuilder.Network));
```
Comment (dangershony) [inline Stratis.Bitcoin.Features.Consensus/ConsensusFeature.cs]: No need to create an instance
``` services.AddSingleton<ICheckpoints, Checkpoints>();```

File: Stratis.Bitcoin.Features.Consensus/PosConsensusValidator.cs [hunk 3]
```diff
@@ -448,7 +450,12 @@ public void CheckAndComputeStake(ContextInformation context)
                 if (prevBlockStake == null)
                     ConsensusErrors.PrevStakeNull.Throw();
 
-                this.stakeValidator.CheckProofOfStake(context, pindexPrev, prevBlockStake, block.Transactions[1], pindex.Header.Bits.ToCompact());
+                // Only do proof of stake validation for blocks after last checkpoint.
+                if (pindex.Height > lastCheckpointHeight)
```
Comment (dangershony) [inline Stratis.Bitcoin.Features.Consensus/PosConsensusValidator.cs]: A side effect is we will not have the stake information of all the other blocks, this needs to be configurable in my opinion.

File: Stratis.Bitcoin.Features.Consensus/ConsensusFeature.cs (hunk 2 above)
Comment (Aprogiena) [reply Stratis.Bitcoin.Features.Consensus/ConsensusFeature.cs]: ok will try

File: Stratis.Bitcoin.Features.Consensus/Checkpoints.cs (hunk 1 above)
Comment (Aprogiena) [reply Stratis.Bitcoin.Features.Consensus/Checkpoints.cs]: ok will try

File: Stratis.Bitcoin.Features.Consensus/Checkpoints.cs [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        };
+
+        /// <summary>Checkpoints for the specific instance of the class and its network.</summary>
+        private readonly Dictionary<int, CheckpointInfo> checkpoints;
+
+        /// <summary>
+        /// Initializes a new instance of the object.
+        /// </summary>
+        /// <param name="network">Specification of the network the node runs on - regtest/testnet/mainnet/stratis test/main.</param>
+        public Checkpoints(Network network)
```
Comment (dangershony) [inline Stratis.Bitcoin.Features.Consensus/Checkpoints.cs]: I guess this is fine for this PR but I would like to see checkpoints registered in the Network class (or consensus)
Then at this point Network would populate the dictionary checkpoints.

Comment (Aprogiena) [reply Stratis.Bitcoin.Features.Consensus/Checkpoints.cs]: It was my original intention until I realized it is part of NBitcoin. So yes, future work can go further towards Network.

File: Stratis.Bitcoin.Features.Consensus/PosConsensusValidator.cs (hunk 3 above)
Comment (Aprogiena) [reply Stratis.Bitcoin.Features.Consensus/PosConsensusValidator.cs]: I am not sure I follow here. I understand the general requirement of having this configurable, this is quite standard requirement for users who do not want to trust devs that much. But I do not understand the side effect you mention.

Comment (Aprogiena) [reply Stratis.Bitcoin.Features.Consensus/PosConsensusValidator.cs]: I will create issue for adding checkpoint config switch.

**Pull request merged**